objects/src/assets/mod.rs

```diff
@@ -0,0 +1,452 @@
+use super::{AccountId, AssetError, Digest, Felt, Hasher, StarkField, Word, ZERO};
+use core::{fmt, ops::Deref};
+
+// ASSET
+// ================================================================================================
+
+/// A fungible or a non-fungible asset.
+///
+/// All assets are encoded using a single word (4 elements) such that it is easy to determine the
+/// type of an asset both inside and outside Miden VM. Specifically:
+/// - The first bit of a fungible asset is always ONE, and the last 32-bits of the 3rd element
+///   are always set to ZEROs.
+/// - The first bit of a non-fungible asset is always ZERO, and the last 32-bits of the 3rd element
+///   are always set to 2^31 (i.e., ONE followed by 31 ZEROs).
+///
+/// The above properties guarantee that there can never be a collision between a fungible and a
+/// non-fungible asset. Collision resistance of both fungible and non-fungible assets is ~110 bits.
+/// However, for fungible assets collision resistance is not important as to get a collision for
+/// fungible assets there must be a collision in account IDs which can be prevented at account
+/// creation time.
+///
+/// The methodology for constructing fungible and non-fungible assets is described below.
+///
+/// # Fungible assets
+/// The first 3 elements of a fungible asset are set to the ID of the faucet which issued the
+/// asset. This guarantees the properties described above (the first bit is ONE, and the 32 least
+/// significant bits of the 3rd element are all ZEROs).
+///
+/// The last element is set to the amount of the asset. This amount cannot be greater than
+/// 2^63 - 1 and thus requires 63-bits to store.
+///
+/// # Non-fungible assets
+/// The 4 elements of non-fungible assets are computed as follows:
+/// - First the asset data is hashed. This compresses an asset of an arbitrary length to 4 field
+///   elements.
+/// - The result of step 1 is hashed together with the ID of the faucet which issued the asset.
```

Comment:
Is there a reason to not use `hash(asset_data || faucet_id)`? Do we want to blind the `asset_data`?

Reply:
Yes, but also I think it will make things a bit easier. i.e., to verify that the asset has been computed correctly, we don't need to hash the full asset data in the VM. Instead, we can hash it outside of the VM and in the VM we just hash data hash and faucet ID.